### DISCUSSION TITLE
How to secure serverless functions??

### DISCUSSION THREAD

**Discussion by lalit2005** [vercel/community#1833] [Help] [answered]
By 'How to secure serverless functions', I mean **How to invoke a serverless function only through my website/domain?**

If I deploy a serverless function, it can be accessed by `/api/filename`.
But this route can be called from any website/domain which can result in depletion of the monthly quota of 100GB-Hrs and also API can be accessed publicly.

So I would like to be able to invoke my function only through my website(eg. `my-website.vercel.app`) and not from any other
domains.

I tried to use CORS after referring to answers on StackOverflow but I was unsuccessful in blocking requests from different origins...😞 

I also tried to find a [parameter in `req`](https://pastebin.com/embed_iframe/bA58McEF) that contained the `origin` of request like in express.js but I could not find one...

I have read in many places that Authentication and Authorization should be used to protect API endpoints, 
In this case will my API be accessible?? (I am new to the jamstack world, please explain like I'm five   )
### Thanks a lot in advance.🙇

**Comment by johnEthicalTechnology** [comment, marked as answer]
@Lalit2005 I've got the same question. Have you come across any answer for this?

**Reply by lalit2005** [reply]
Hello @johnEthicalTechnology, I found out later that, serverless functions cannot be called from other domains because Vercel protects serverless functions from being called from another domain with the help of CORS.
But the serverless functions can be invoked from any node environment 😞 
So, if a serverless function has to be protected JWT tokens can be used..
Thus, I think that authentication and authorization are the only way to go

Please correct me if I'm wrong 🙏

**Reply by johnEthicalTechnology** [reply]
@Lalit2005 authentication and authorization is one way to do it.

The server that is calling my serverless function doesn't have the capability for a authentication and authorization workflow.

The option I'm going with is to create a HMAC-SHA256 string with key and data stored securely on both the calling server and the serverless function. The serverless function will test the incoming request for that string and if it's not the same it'll fail.

That's the easiest way I can think about securing the serverless function within the limitation that I have.

**Reply by lalit2005** [reply]
Yeah that's possible too!
Hope this helps → https://github.com/baulml/nextjs-cron#authorization-flow

**Reply by AlessandroVol23** [reply]
Hi you two. I am facing the same issue currently but I am not sure if I understand the answer 100 % correct.

>  I found out later that, serverless functions cannot be called from other domains because Vercel

I've read that already as well. But if I make a curl on `<DEPLOYMENT_URL/api/apiName` I get a response and no error. 

Do I need a JWT authentication to secure my API? I know that in AWS it is working either with API-Keys or IAM authentication. Is something similar available for Vercel? API key via environment variables doesn't make sense I think if you need to communicate from the frontend.

Would be awesome if you have a tip on how you solved this one. Thanks a lot! :)

**Reply by lalit2005** [reply]
Hey @AlessandroVol23, I finally went with JWT authorization as you mentioned. JWT combined with [rate limiting](https://github.com/vercel/next.js/tree/canary/examples/api-routes-rate-limit) served my purpose

> I've read that already as well. But if I make a curl on <DEPLOYMENT_URL/api/apiName I get a response and no error.

CORS only protects the api route from being called from a browser environment afaik  
When using JWT or session tokens, API routes can be protected from requests made from non-browser environments

If you are using [Auth0](https://auth0.github.io/nextjs-auth0/modules/helpers_with_api_auth_required.html) or [Clerk](https://docs.clerk.dev/get-started/nextjs-api), they provide useful functions such as `withApiAuthRequired` and `requireSession` to authorize API endpoints in Next.js (they use JWT under the hood to authorize requests). I believe [next-auth](http://next-auth.js.org/) also has similar methods.

If you do not use any service such as Auth0 or Clerk, you can just use the `jwt.verify()` and `jwt.sign()` methods from the `jsonwebtoken` library for encoding and decoding tokens. 

And additionally rate limiting can be done with the help of [Next.js 12 middlewares](https://github.com/vercel/examples/tree/main/edge-functions/api-rate-limit) or with the help of [lru-cache](https://github.com/vercel/next.js/tree/canary/examples/api-routes-rate-limit).

---

> API key via environment variables doesn't make sense I think if you need to communicate from the frontend

Yeah, you can store a single API key in environment variables if you want to invoke the function from a different non-browser environment securely. For eg. if you would like to run a cron job and invoke the function periodically, you can ping it with a service such as Uptime robot with the API key in query params. And before executing the main logic, you can check for the correct API key in query params👇

```js
export default function handler(req, res) {
  if (req.query.SECRET_KEY !== process.env.SECRET_KEY) {
      res.status(401).json({
        status: 401
        message: 'Unauthorized'
      })
      return;
  }

  // main logic of the function

}
```

I hope this answers your question :)

**Reply by sgbaird** [reply]
@Lalit2005, to clarify, is putting the API key in the query parameters secure against others seeing the unencrypted API key?

**Reply by ZeldOcarina** [reply]
@sgbaird it's not, it should go on the Headers.
